File: modules/@angular/core/test/reflection/reflector_common.ts
import {makeDecorator, makeParamDecorator, makePropDecorator} from '@angular/core/src/util/decorators';

export class ClassDecoratorMeta {
  constructor(public value: any /** TODO #9100 */) {}
}

export class ParamDecoratorMeta {
  constructor(public value: any /** TODO #9100 */) {}
}

export class PropDecoratorMeta {
  constructor(public value: any /** TODO #9100 */) {}
}

export function classDecorator(value: any /** TODO #9100 */) {
  return new ClassDecoratorMeta(value);
}

export function paramDecorator(value: any /** TODO #9100 */) {
  return new ParamDecoratorMeta(value);
}

export function propDecorator(value: any /** TODO #9100 */) {
  return new PropDecoratorMeta(value);
}

/** @Annotation */ export var ClassDecorator = makeDecorator(ClassDecoratorMeta);
/** @Annotation */ export var ParamDecorator = makeParamDecorator(ParamDecoratorMeta);
/** @Annotation */ export var PropDecorator = makePropDecorator(PropDecoratorMeta);

// used only in Dart
export class HasGetterAndSetterDecorators {}

File: modules/@angular/core/test/reflection/reflector_common.dart
class ClassDecorator {
  final dynamic value;

  const ClassDecorator(this.value);
}

class ParamDecorator {
  final dynamic value;

  const ParamDecorator(this.value);
}

class PropDecorator {
  final dynamic value;

  const PropDecorator(this.value);
}

ClassDecorator classDecorator(value) {
  return new ClassDecorator(value);
}

ParamDecorator paramDecorator(value) {
  return new ParamDecorator(value);
}

PropDecorator propDecorator(value) {
  return new PropDecorator(value);
}

class HasGetterAndSetterDecorators {
  @PropDecorator("get") get a {}
  @PropDecorator("set") set a(v) {}
}
